man/runsvstat.8
.TH runsvstat 8
.SH NAME
runsvstat \- prints the status of services monitored by
.BR runsv (8).
.SH SYNOPSIS
.B runsvstat
[
.B \-l
]
.I services
.SH DESCRIPTION
.I services
consists of one or more arguments, each argument naming a directory.
.P
.B runsvstat
checks for each directory if
.BR runsv (8)
is successfully running this
.IR service ,
and if so, prints one human-readable line reporting the status of this
.IR service .
Optionally
.BR runsvstat (8)
prints a second line, also in human-readable format, reporting the status
of the corresponding
.I log
.IR service .
.SH OPTIONS
.TP
.B \-l
log service.
Check for each
.I service
if there exists a corresponding
.I log
.IR service .
If so, print an additional, indented, line reporting the status of the
.I log
.IR service .
.SH EXIT CODES
.B runsvstat
exits 0, if the states of all
.I services
were reported successfully.
.P
For each
.I service
that caused an error (e.g. the directory is not controlled by a
.BR runsv (8)
process),
.B runsvstat
increases the exit code by one and exits non zero.
The maximum is 100.
.P
.B runsvstat
exits 111 on error.
.SH SEE ALSO
sv(8),
runsv(8),
chpst(8),
svlogd(8),
runit(8),
runit-init(8),
runsvdir(8),
runsvchdir(8),
runsvctrl(8)
.P
http://smarden.org/runit/
.SH AUTHOR
Gerrit Pape <pape@smarden.org>


package/CHANGES
runit 1.3.2
Sun, 18 Dec 2005 11:22:04 +0000
  * svlogd.c: don't print extra newlines to the log if additionally writing
    to the network through udp (thx Andras Korn,
    http://bugs.debian.org/339030).
  * runsvdir.c: also process svdir changes if mtime of svdir has changed into
    the past (thx Martin Dickopp).
  * doc/index.html: update description; Please contact this list and not me
    privately.
  * doc/replaceinit.html, doc/runscripts.html: minor.
  * man/chpst.8: -U doesn't support a list of groups.
  * man/svlogd.8: update PATTERN MATCHING section, add example (based on
    sugesstions from Joshua N Pritikin).
  * man/*8: update SEE ALSO sections: runsvctrl, runsvstat, svwaitdown,
    svwaitup will become obsolete; remove references to man pages for
    programs from the daemontools package.

runit 1.3.1
Wed, 24 Aug 2005 20:14:17 +0000
  * doc/runscripts.html: typo; openssh needs absolute path (thx Kevin Berry);
    exec into dhcpcd.
  * uidgid.c, uidgid.h, chpst.c: support colon-separated list of groups.
  * sv.c: utilize optional ./check script in service directory to check for
    availablity of service.
  * runsv.c: wait_pid() might be interupted.
  * man/chpst.8, man/sv.8: adapt.

runit 1.3.0
Sun, 24 Jul 2005 16:50:55 +0000
  * man/runsv.8: typo; no longer document the e control character; clarify
    custom control on d and x.
  * doc/useinit.html: how to use runit with macosx 10.4 (thx Lars Uffmann).
  * etc/macosx/org.smarden.runit.plist: new; launchd plist file for
    runsvdir-start (thx Lars Uffmann).
  * doc/runscripts.html: don't use absolute pathnames for programs in $PATH;
    add contributed run scripts (thx Sascha Huedepohl); typos.
  * runsvchdir.c, runsvstat.c: optimize output buffer handling.
  * runsvctrl.c: write x control character to runsv on e.
  * svlogd.c: write to log dir after possibly writing through udp (fixes
    line break issue).
  * runsv.c, runsv.check: optimize output buffer handling; porperly run
    custom t, d, x, k, when stopping service; no longer support e control
    character.
  * sv.c: new sv program to control services, optionally can be linked to
    /etc/init.d/ as lsb "init script" command line interface.
  * Makefile, TARGETS: adapt.
  * chpst.c: fix pointer types.
  * man/sv.8: new.
  * package/TODO: new.

runit 1.2.3
Mon, 18 Apr 2005 17:08:35 +0000
  * doc/runscripts.html: add contributed run scripts (thx Kevin Berry);
    remove mysql run scripts.
  * svlogd.dist, svwaitdown.dist, svwaitup.dist: fix program selftest's
    false alarms (thx Ian Stokes-Rees).

runit 1.2.2
Sun, 03 Apr 2005 09:21:52 +0000
  * doc/index.html, doc/replaceinit.html: no longer refer to sysvinit's
    shutdown program.
  * man/svlogd.8: svlogd doesn't decrease the number of log files in a log
    directory; with n0 don't remove any log files.
  * svlogd.c: properly prefix log messages written to standard error with
    optional timestamp.
  * runsvctrl.c, svwaitdown.c, svwaitup.c: supervise not running -> runsv
    not running.
  * package/README: 2005.

runit 1.2.1
Sat, 15 Jan 2005 11:57:46 +0000
  * chpst.c: typo.
  * svwaitdown.c: minimum timeout is 1 second; service directories no longer
    must start with /.
  * svwaitup.c: minimum uptime of services is 1 second; service directories
    no longer must start with /; print number of seconds a service is up if
    verbose.
  * man/svwaitdown.8, man/svwaitup.8: adapt.
  * doc/runscripts.html: add contributed run scripts (thx Stefan Karrmann,
    Kevin Berry, Karl Chen).
  * runsv.c: typos; bugfix: run optional control/x when receiving SIGTERM
    (thx Vladimir Smelhaus); minor.
  * man/runsvctrl.8: minor.
  * package/COPYING: adapt, 2005.

runit 1.2.0
Fri, 17 Dec 2004 21:08:50 +0000
  * doc/upgrade.html: typo.
  * runsv.c: support custom control commands through control/ directory,
    optionally switch off sending signal; don't update timestamp in status
    when running ./finish; only sleep for a second if ./run has been
    restarted within a second, and after updating status to down, normally
    up, want up.
  * man/runsv.8: document custom control commands.
  * runsv.check, runsv.dist: check custom control commands.
  * doc/runscripts.html: add contributed run scripts (thx Stefan Karrmann,
    Kevin Berry).

runit 1.1.0
Sat, 06 Nov 2004 17:21:11 +0000
  * svlogd.c: new config option t timeout (thx Enrico Scholz); config
    options e and E select and deselect lines for stderr respectively (thx
    Vladimir Smelhaus); new config option N.
  * man/svlogd.8: adapt.
  * runsv.c: on commands down and exit send CONT after TERM.
  * man/runsv.8: adapt.
  * etc/2: use -P option to runsvdir.
  * src/svlogd.check: add check for t config option.
  * chpst.c: new option -n: adjust nice level.
  * man/chpst.8: adapt.

runit 1.0.5
Tue, 21 Sep 2004 18:18:14 +0000
  * svlogd.c: fix line buffer handling for pattern matching (thx Enrico
    Scholz); properly retry writing to current on error; minor.
  * doc/index.html: wording; add link to useinit.html; remove link to runit
    on linuxfromscratch.
  * doc/runscripts.html: add hint, some getties need to be run in a new
    session/process group (thx Dan Melomedman).
  * man/svlogd.8: minor fixes.
  * man/svwaitdown.8: remove hint about -x option.
  * package/check, package/compile, package/install, package/install-man,
    package/upgrade: use set -e instead of #!/bin/sh -e.
  * runsvdir.c: new option -P, run each runsv process in a new session and
    process group (thx Charlie Brady).
  * man/runsvdir.8: adapt; typo.

runit 1.0.4
Sun, 01 Aug 2004 18:29:36 +0000
  * doc/runlevels.html, doc/upgrade.html, doc/useinit.html, man/runsvdir.8,
    man/runsvchdir.8: minor changes in wording, notations.
  * runsvdir.c: check services dir again after one second until it's
    unchanged (thx Eric Lammerts); sleep one second before scanning if
    services dir mtime is now; loop while chdir to starting dir fails;
    rename log, loglen to rplog, rploglen.

src/runsvstat.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "strerr.h"
#include "error.h"
#include "sgetopt.h"
#include "open.h"
#include "buffer.h"
#include "tai.h"
#include "fmt.h"

#define USAGE " [ -l ] service ..."

#define VERSION "$Id: runsvstat.c,v 1.8 2005/07/11 10:13:53 pape Exp $"

#define FATAL "runsvstat: fatal: "
#define WARNING "runsvstat: warning: "

const char *progname;
unsigned int rc =0;
struct stat s;
int showlog =0;

void usage() { strerr_die4x(1, "usage: ", progname, USAGE, "\n"); }

void fatal(char *m1) { strerr_die3sys(111, FATAL, m1, ": "); }
void warn(char *m1, char *m2) {
  rc++;
  strerr_warn5(WARNING, m1, ": ", m2, ": ", &strerr_sys);
}
void warnx(char *m1, char *m2) {
  rc++;
  strerr_warn4(WARNING, m1, ": ", m2, 0);
}

int show_status(char *name) {
  char status[20];
  int pid;
  int fd;
  int normallyup =0;
  char sulong[FMT_ULONG];
  struct tai when;
  struct tai now;

  if (stat("down", &s) == -1) {
    if (errno != error_noent) {
      warn(name, "unable to stat down");
      return(-1);
    }
    normallyup = 1;
  }
  if ((fd =open_write("supervise/ok")) == -1) {
    if (errno == error_nodevice)
      warnx(name, "runsv not running.");
    else
      warn(name, "unable to open supervise/ok");
    return(-1);
  }
  close(fd);
  if ((fd =open_read("supervise/status")) == -1) {
    warn(name, "unable to open supervise/status");
    return(-1);
  }
  switch(read(fd, status, 20)) {
  case 20: break;
  case -1:
    warn(name, "unable to read supervise/status");
    return(-1);
  default:
    warnx(name, "unable to read supervise/status: bad format.");
    return(-1);
  }
  pid =(unsigned char) status[15];
  pid <<=8; pid +=(unsigned char)status[14];
  pid <<=8; pid +=(unsigned char)status[13];
  pid <<=8; pid +=(unsigned char)status[12];

  tai_unpack(status,&when);
  tai_now(&now);
  if (tai_less(&now,&when)) when =now;
  tai_sub(&when,&now,&when);

  buffer_puts(buffer_1, name);
  buffer_puts(buffer_1, ": ");
  if (pid) {
    switch (status[19]) {
    case 1: buffer_puts(buffer_1, "run "); break;
    case 2: buffer_puts(buffer_1, "finish "); break;
    }
    buffer_puts(buffer_1, "(pid ");
    buffer_put(buffer_1, sulong, fmt_ulong(sulong, pid));
    buffer_puts(buffer_1, ") ");
  }
  else
    buffer_puts(buffer_1, "down ");
  buffer_put(buffer_1, sulong, fmt_ulong(sulong, tai_approx(&when)));
  buffer_puts(buffer_1, " seconds");
  if (pid && !normallyup) buffer_puts(buffer_1,", normally down");
  if (!pid && normallyup) buffer_puts(buffer_1,", normally up");
  if (pid && status[16]) buffer_puts(buffer_1,", paused");
  if (!pid && (status[17] == 'u')) buffer_puts(buffer_1,", want up");
  if (pid && (status[17] == 'd')) buffer_puts(buffer_1,", want down");
  if (pid && status[18]) buffer_puts(buffer_1, ", got TERM");
  /* buffer_putsflush(buffer_1, "\n"); */
  return(1);
}

int main(int argc, char **argv) {
  int opt;
  int curdir;
  char **dir;

  progname =*argv;

  while ((opt =getopt(argc, (const char * const *)argv, "lV")) != opteof) {
    switch(opt) {
    case 'l':
      showlog =1;
      break;
    case 'V':
      strerr_warn1(VERSION, 0);
    case '?':
      usage();
    }
  }
  argv +=optind;

  dir =argv;
  if (! dir || ! *dir) usage();

  if ((curdir =open_read(".")) == -1) {
    rc =100;
    fatal("unable to open current directory");
  }
  for (; dir && *dir; dir++) {
    if (chdir(*dir) == -1) {
      warn(*dir, "unable to change directory");
      continue;
    }
    if (show_status(*dir) == 1) {
      if (showlog) {
	if (stat("log", &s) == -1) {
	  if (errno != error_noent)
	    warn("unable to stat()", "./log");
	}
	else {
	  if (! S_ISDIR(s.st_mode))
	    warnx("./log", "not a directory.");
	  else {
	    if (chdir("log") == -1) {
	      warn(*dir, "unable to change directory");
	      continue;
	    }
	    show_status("\n  log");
	  }
	}
      }
      buffer_puts(buffer_1, "\n"); buffer_flush(buffer_1);
    }
    if (fchdir(curdir) == -1) {
      rc =100;
      fatal("unable to change directory");
    }
  }
  if (rc > 100) rc =100;
  _exit(rc);
}
